### ISSUE TITLE
Einops for batched computations

### ISSUE THREAD

**Issue by enver1323** [patrick-kidger/equinox#636]
Hi, I really enjoy using equinox, but sometimes, when I am trying to process high-dimensional data with `jax.vmap` it can be a quite slow. Therefore, I was wondering, if it is feasible to change some operations of modules with einops operations. I am not sure, if this can be adopted, and it does not go against framework philosophy, but just a kind idea of improving performance based on my personal experience.
**Current implementation**:
`nn.Linear.__call__`:
```
x = self.weight @ x
```
In code:
```
model = nn.Linear(336, 512)
jax.vmap(model)(x, y)
```
Proposed implementation:
`nn.Linear.__call__`:
```
x = einsum(self.weight, x, 'a b, ... b -> ... a')
```
In code:
```
model = nn.Linear(336, 512)
model(x, y)
```
Although without `eqx.filter_jit`, the performance is almost identical, I experienced around x10 speed boost with compilation.

**Comment by patrick-kidger** [comment]
Hmm, this is unfortunate! Thank you for the reproduction. This seems to be a JAX/XLA performance bug, so I've opened this at https://github.com/google/jax/issues/19327.

Sadly XLA seems to have a few of these. I'm not sure if there's much we can do from the Equinox side. In particular, whilst we might get a speedup by changing things here, it's not clear to me whether such a change might introduce a slowdown elsewhere.

FWIW, at least in my own testing, you can work around this by replacing the double-vmap with a single vmap, and `jnp.reshape`ing to flatten the two leading dimensions together before hand, and then `jnp.reshape`ing again afterwards to split them apart. This gives good performance once again.

**Comment by dlwh** [comment]
just throwing in my two cents here: I've run into a lot of vmap performance issues, and if you want every ounce of performance out of JAX you should avoid vmap. It's especially bad with RNG, though it's being worked on.

**Comment by enver1323** [comment]
Thank you, @patrick-kidger for your kind reply and openning another issue.
> In particular, whilst we might get a speedup by changing things here, it's not clear to me whether such a change might introduce a slowdown elsewhere.

Ah, yes, this was one of my concerns while trying to use `einsum` projectwide instead of `vmap` too.

> FWIW, at least in my own testing, you can work around this by replacing the double-vmap with a single vmap, and `jnp.reshape`ing to flatten the two leading dimensions together before hand, and then `jnp.reshape`ing again afterwards to split them apart. This gives good performance once again.

Thank you for your suggestion. I also tried to do this at first, but for some reason, I sometimes get even better performance with `einsum`, when testing on GPU, if compilation time is not considered.

**Comment by enver1323** [comment]
> just throwing in my two cents here: I've run into a lot of vmap performance issues, and if you want every ounce of performance out of JAX you should avoid vmap. It's especially bad with RNG, though it's being worked on.

Yes, unfortunately, I have the same opinion, @dlwh. I also felt like using `vmap` is much more memory-consuming
